[ERROR] appsignal: Error in minutely probe 'sidekiq': ERR unknown command `connection`, with args beginning with:

If your bug is about how AppSignal works in your app specifically we recommend you contact us at support@appsignal.com with your bug report instead.

---

## Describe the bug
Just getting spam that says "appsignal: Error in minutely probe 'sidekiq': ERR unknown command `connection`, with args beginning with:  " over and over again

## To Reproduce

Steps to reproduce the behavior:

- Using AppSignal for Ruby gem version 3.1.5
- sidekiq 7.0, rails 7.0.4

No code to run, fails non-stop

@Aeramor For me it fails even sooner on minutely probe start up, with this error:

```
[2022-11-07T13:49:00 (process) #230][ERROR] Error while initializing minutely probe 'sidekiq': uninitialized constant Redis

        Gem::Version.new(::Redis::VERSION) >= Gem::Version.new("3.3.5")
                         ^^^^^^^
[2022-11-07T13:49:00 (process) #230][DEBUG] /app/vendor/bundle/ruby/3.1.0/gems/appsignal-3.1.5/lib/appsignal/probes/sidekiq.rb:11:in `dependencies_present?'
```

Looks like sidekiq changed their redis dependency to redis-client, named `RedisClient`, hence the name error.

I'm not sure I'm reproducing this properly yet. Or do you have the `redis` gem in your app's Gemfile as well?

If I remove the version check the failed in my last comment I get a similar error as originally reported:

```
[2022-11-07T14:26:22 (process) #435][ERROR] Error in minutely probe 'sidekiq': ERR unknown command `connection`, with args beginning with:
[2022-11-07T14:26:22 (process) #435][DEBUG] /app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client/connection_mixin.rb:22:in `call'
/app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client.rb:200:in `block (2 levels) in call'
/app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client/middlewares.rb:16:in `call'
/app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client.rb:199:in `block in call'
/app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client.rb:616:in `ensure_connected'
/app/vendor/bundle/ruby/3.1.0/gems/redis-client-0.11.1/lib/redis_client.rb:198:in `call'
/app/vendor/bundle/ruby/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq/redis_client_adapter.rb:25:in `method_missing'
/integration/lib/appsignal/probes/sidekiq.rb:92:in `block in hostname'
/app/vendor/bundle/ruby/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq/config.rb:156:in `block in redis'
/app/vendor/bundle/ruby/3.1.0/gems/connection_pool-2.3.0/lib/connection_pool.rb:65:in `block (2 levels) in with'
/app/vendor/bundle/ruby/3.1.0/gems/connection_pool-2.3.0/lib/connection_pool.rb:64:in `handle_interrupt'
/app/vendor/bundle/ruby/3.1.0/gems/connection_pool-2.3.0/lib/connection_pool.rb:64:in `block in with'
/app/vendor/bundle/ruby/3.1.0/gems/connection_pool-2.3.0/lib/connection_pool.rb:61:in `handle_interrupt'
/app/vendor/bundle/ruby/3.1.0/gems/connection_pool-2.3.0/lib/connection_pool.rb:61:in `with'
/app/vendor/bundle/ruby/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq/config.rb:153:in `redis'
/app/vendor/bundle/ruby/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq.rb:73:in `redis'
/integration/lib/appsignal/probes/sidekiq.rb:92:in `hostname'
/integration/lib/appsignal/probes/sidekiq.rb:78:in `gauge'
/integration/lib/appsignal/probes/sidekiq.rb:52:in `track_stats'
/integration/lib/appsignal/probes/sidekiq.rb:31:in `call'
/integration/lib/appsignal/minutely.rb:140:in `block (3 levels) in start'
/integration/lib/appsignal/minutely.rb:137:in `each'
/integration/lib/appsignal/minutely.rb:137:in `block (2 levels) in start'
/integration/lib/appsignal/minutely.rb:134:in `loop'
/integration/lib/appsignal/minutely.rb:134:in `block in start'
```

Looks like we'll have to update the probe to work with the redis and redis-client gems.

--- a/.changesets/support-sidekiq-7-in-the-sidekiq-minutely-probe.md
+++ b/.changesets/support-sidekiq-7-in-the-sidekiq-minutely-probe.md
@@ -3,4 +3,4 @@ bump: "patch"
 type: "fix"
 ---
 
-Support Sidekiq 7 in the Sidekiq minutely probe. Fixes the "unknown command `connection`" error logged to `appsignal.log`.
+Support Sidekiq 7 in the Sidekiq minutely probe. It will now report metrics to Sidekiq magic dashboard for Sidekiq version 7 and newer.

--- a/.changesets/support-sidekiq-7-in-the-sidekiq-minutely-probe.md
+++ b/.changesets/support-sidekiq-7-in-the-sidekiq-minutely-probe.md
@@ -1,0 +1,6 @@
+---
+bump: "patch"
+type: "fix"
+---
+
+Support Sidekiq 7 in the Sidekiq minutely probe. Fixes the "unknown command `connection`" error logged to `appsignal.log`.

--- a/lib/appsignal/probes/sidekiq.rb
+++ b/lib/appsignal/probes/sidekiq.rb
@@ -6,8 +6,15 @@ module Appsignal
       # @api private
       attr_reader :config
 
+      def self.sidekiq7_and_greater?
+        Gem::Version.new(::Sidekiq::VERSION) >= Gem::Version.new("7.0.0")
+      end
+
       # @api private
       def self.dependencies_present?
+        return true if sidekiq7_and_greater?
+        return unless defined?(::Redis::VERSION) # Sidekiq <= 6
+
         Gem::Version.new(::Redis::VERSION) >= Gem::Version.new("3.3.5")
       end
 
@@ -31,8 +38,15 @@ module Appsignal
       attr_reader :cache
 
       def track_redis_info
-        return unless ::Sidekiq.respond_to?(:redis_info)
-        redis_info = ::Sidekiq.redis_info
+        redis_info = nil
+        if self.class.sidekiq7_and_greater?
+          # Sidekiq >= 7 using the redis-client gem
+          ::Sidekiq.redis { |c| redis_info = c.info }
+        elsif ::Sidekiq.respond_to?(:redis_info)
+          # Sidekiq <= 6 using the redis gem
+          redis_info = ::Sidekiq.redis_info
+        end
+        return unless redis_info
 
         gauge "connection_count", redis_info.fetch("connected_clients")
         gauge "memory_usage", redis_info.fetch("used_memory")
@@ -82,7 +96,14 @@ module Appsignal
         end
 
         host = nil
-        ::Sidekiq.redis { |c| host = c.connection[:host] }
+        ::Sidekiq.redis do |c|
+          host =
+            if c.respond_to? :connection
+              c.connection[:host] # Sidekiq <= 6
+            else
+              c.config.host # Sidekiq >= 7
+            end
+        end
         Appsignal.logger.debug "Sidekiq probe: Using Redis server hostname " \
           "#{host.inspect} as hostname"
         @hostname = host

--- a/lib/appsignal/probes/sidekiq.rb
+++ b/lib/appsignal/probes/sidekiq.rb
@@ -3,6 +3,38 @@ module Appsignal
     class SidekiqProbe
       include Helpers
 
+      class Sidekiq7Adapter
+        def self.redis_info
+          redis_info = nil
+          ::Sidekiq.redis { |c| redis_info = c.info }
+          redis_info
+        end
+
+        def self.hostname
+          host = nil
+          ::Sidekiq.redis do |c|
+            host = c.config.host
+          end
+          host
+        end
+      end
+
+      class Sidekiq6Adapter
+        def self.redis_info
+          return unless ::Sidekiq.respond_to?(:redis_info)
+
+          ::Sidekiq.redis_info
+        end
+
+        def self.hostname
+          host = nil
+          ::Sidekiq.redis do |c|
+            host = c.connection[:host] if c.respond_to? :connection
+          end
+          host
+        end
+      end
+
       # @api private
       attr_reader :config
 
@@ -21,6 +53,9 @@ module Appsignal
       def initialize(config = {})
         @config = config
         @cache = {}
+        is_sidekiq7 = self.class.sidekiq7_and_greater?
+        @adapter = is_sidekiq7 ? Sidekiq7Adapter : Sidekiq6Adapter
+
         config_string = " with config: #{config}" unless config.empty?
         Appsignal.logger.debug("Initializing Sidekiq probe#{config_string}")
         require "sidekiq/api"
@@ -35,17 +70,10 @@ module Appsignal
 
       private
 
-      attr_reader :cache
+      attr_reader :adapter, :cache
 
       def track_redis_info
-        redis_info = nil
-        if self.class.sidekiq7_and_greater?
-          # Sidekiq >= 7 using the redis-client gem
-          ::Sidekiq.redis { |c| redis_info = c.info }
-        elsif ::Sidekiq.respond_to?(:redis_info)
-          # Sidekiq <= 6 using the redis gem
-          redis_info = ::Sidekiq.redis_info
-        end
+        redis_info = adapter.redis_info
         return unless redis_info
 
         gauge "connection_count", redis_info.fetch("connected_clients")
@@ -95,15 +123,7 @@ module Appsignal
           return @hostname
         end
 
-        host = nil
-        ::Sidekiq.redis do |c|
-          host =
-            if c.respond_to? :connection
-              c.connection[:host] # Sidekiq <= 6
-            else
-              c.config.host # Sidekiq >= 7
-            end
-        end
+        host = adapter.hostname
         Appsignal.logger.debug "Sidekiq probe: Using Redis server hostname " \
           "#{host.inspect} as hostname"
         @hostname = host

--- a/spec/lib/appsignal/probes/sidekiq_spec.rb
+++ b/spec/lib/appsignal/probes/sidekiq_spec.rb
@@ -136,6 +136,8 @@ describe Appsignal::Probes::SidekiqProbe do
     def with_sidekiq7!
       stub_const("Sidekiq", Sidekiq7Mock)
     end
+    # Version not relevant, but requires any version for tests
+    alias_method :with_sidekiq!, :with_sidekiq7!
 
     def with_sidekiq6!
       stub_const("Sidekiq", Sidekiq6Mock)
@@ -175,7 +177,7 @@ describe Appsignal::Probes::SidekiqProbe do
     end
 
     it "loads Sidekiq::API" do
-      with_sidekiq7! # Version irrelevant, but requires any version
+      with_sidekiq!
       # Hide the Sidekiq constant if it was already loaded. It will be
       # redefined by loading "sidekiq/api" in the probe.
       hide_const "Sidekiq::Stats"
@@ -186,7 +188,7 @@ describe Appsignal::Probes::SidekiqProbe do
     end
 
     it "logs config on initialize" do
-      with_sidekiq7! # Version irrelevant, but requires any version
+      with_sidekiq!
       log = capture_logs { probe }
       expect(log).to contains_log(:debug, "Initializing Sidekiq probe\n")
     end
@@ -281,7 +283,7 @@ describe Appsignal::Probes::SidekiqProbe do
       let(:probe) { described_class.new(:hostname => redis_hostname) }
 
       it "uses the redis hostname for the hostname tag" do
-        with_sidekiq7! # Version irrelevant, but requires any version
+        with_sidekiq!
 
         allow(Appsignal).to receive(:set_gauge).and_call_original
         log = capture_logs { probe }

--- a/spec/lib/appsignal/probes/sidekiq_spec.rb
+++ b/spec/lib/appsignal/probes/sidekiq_spec.rb
@@ -7,7 +7,103 @@ describe Appsignal::Probes::SidekiqProbe do
     let(:expected_default_tags) { { :hostname => "localhost" } }
     before do
       Appsignal.config = project_fixture_config
-      module SidekiqMock
+
+      class SidekiqStats
+        class << self
+          attr_reader :calls
+
+          def count_call
+            @calls ||= -1
+            @calls += 1
+          end
+        end
+
+        def workers_size
+          # First method called, so count it towards a call
+          self.class.count_call
+          24
+        end
+
+        def processes_size
+          25
+        end
+
+        # Return two different values for two separate calls.
+        # This allows us to test the delta of the value send as a gauge.
+        def processed
+          [10, 15][self.class.calls]
+        end
+
+        # Return two different values for two separate calls.
+        # This allows us to test the delta of the value send as a gauge.
+        def failed
+          [10, 13][self.class.calls]
+        end
+
+        def retry_size
+          12
+        end
+
+        # Return two different values for two separate calls.
+        # This allows us to test the delta of the value send as a gauge.
+        def dead_size
+          [10, 12][self.class.calls]
+        end
+
+        def scheduled_size
+          14
+        end
+
+        def enqueued
+          15
+        end
+      end
+
+      class SidekiqQueue
+        Queue = Struct.new(:name, :size, :latency)
+
+        def self.all
+          [
+            Queue.new("default", 10, 12),
+            Queue.new("critical", 1, 2)
+          ]
+        end
+      end
+
+      module Sidekiq7Mock
+        VERSION = "7.0.0".freeze
+
+        def self.redis
+          yield Client.new
+        end
+
+        class Client
+          def config
+            Config.new
+          end
+
+          def info
+            {
+              "connected_clients" => 2,
+              "used_memory" => 1024,
+              "used_memory_rss" => 512
+            }
+          end
+        end
+
+        class Config
+          def host
+            "localhost"
+          end
+        end
+
+        Stats = ::SidekiqStats
+        Queue = ::SidekiqQueue
+      end
+
+      module Sidekiq6Mock
+        VERSION = "6.9.9".freeze
+
         def self.redis_info
           {
             "connected_clients" => 2,
@@ -26,95 +122,60 @@ describe Appsignal::Probes::SidekiqProbe do
           end
         end
 
-        class Stats
-          class << self
-            attr_reader :calls
-
-            def count_call
-              @calls ||= -1
-              @calls += 1
-            end
-          end
-
-          def workers_size
-            # First method called, so count it towards a call
-            self.class.count_call
-            24
-          end
-
-          def processes_size
-            25
-          end
-
-          # Return two different values for two separate calls.
-          # This allows us to test the delta of the value send as a gauge.
-          def processed
-            [10, 15][self.class.calls]
-          end
-
-          # Return two different values for two separate calls.
-          # This allows us to test the delta of the value send as a gauge.
-          def failed
-            [10, 13][self.class.calls]
-          end
-
-          def retry_size
-            12
-          end
-
-          # Return two different values for two separate calls.
-          # This allows us to test the delta of the value send as a gauge.
-          def dead_size
-            [10, 12][self.class.calls]
-          end
-
-          def scheduled_size
-            14
-          end
-
-          def enqueued
-            15
-          end
-        end
-
-        class Queue
-          Queue = Struct.new(:name, :size, :latency)
-
-          def self.all
-            [
-              Queue.new("default", 10, 12),
-              Queue.new("critical", 1, 2)
-            ]
-          end
-        end
+        Stats = ::SidekiqStats
+        Queue = ::SidekiqQueue
       end
-      stub_const("Sidekiq", SidekiqMock)
     end
-    after { Object.send(:remove_const, :SidekiqMock) }
+    after do
+      Object.send(:remove_const, :SidekiqStats)
+      Object.send(:remove_const, :SidekiqQueue)
+      Object.send(:remove_const, :Sidekiq6Mock)
+      Object.send(:remove_const, :Sidekiq7Mock)
+    end
+
+    def with_sidekiq7!
+      stub_const("Sidekiq", Sidekiq7Mock)
+    end
+
+    def with_sidekiq6!
+      stub_const("Sidekiq", Sidekiq6Mock)
+    end
 
     describe ".dependencies_present?" do
-      before do
-        stub_const("Redis::VERSION", version)
-      end
+      context "when Sidekiq 7" do
+        before { with_sidekiq7! }
 
-      context "when Redis version is < 3.3.5" do
-        let(:version) { "3.3.4" }
-
-        it "does not start probe" do
-          expect(described_class.dependencies_present?).to be_falsy
+        it "starts the probe" do
+          expect(described_class.dependencies_present?).to be_truthy
         end
       end
 
-      context "when Redis version is >= 3.3.5" do
-        let(:version) { "3.3.5" }
+      context "when Sidekiq 6" do
+        before do
+          with_sidekiq6!
+          stub_const("Redis::VERSION", version)
+        end
 
-        it "does not start probe" do
-          expect(described_class.dependencies_present?).to be_truthy
+        context "when Redis version is < 3.3.5" do
+          let(:version) { "3.3.4" }
+
+          it "does not start probe" do
+            expect(described_class.dependencies_present?).to be_falsy
+          end
+        end
+
+        context "when Redis version is >= 3.3.5" do
+          let(:version) { "3.3.5" }
+
+          it "starts the probe" do
+            expect(described_class.dependencies_present?).to be_truthy
+          end
         end
       end
     end
 
     it "loads Sidekiq::API" do
+      with_sidekiq7! # Version irrelevant, but requires any version
       # Hide the Sidekiq constant if it was already loaded. It will be
       # redefined by loading "sidekiq/api" in the probe.
       hide_const "Sidekiq::Stats"
@@ -125,52 +186,93 @@ describe Appsignal::Probes::SidekiqProbe do
     end
 
     it "logs config on initialize" do
+      with_sidekiq7! # Version irrelevant, but requires any version
       log = capture_logs { probe }
       expect(log).to contains_log(:debug, "Initializing Sidekiq probe\n")
     end
 
-    it "logs used hostname on call once" do
-      log = capture_logs { probe.call }
-      expect(log).to contains_log(
-        :debug,
-        %(Sidekiq probe: Using Redis server hostname "localhost" as hostname)
-      )
-      log = capture_logs { probe.call }
-      # Match more logs with incompelete message
-      expect(log).to_not contains_log(:debug, %(Sidekiq probe: ))
-    end
+    context "with Sidekiq 7" do
+      before { with_sidekiq7! }
 
-    it "collects custom metrics" do
-      expect_gauge("worker_count", 24).twice
-      expect_gauge("process_count", 25).twice
-      expect_gauge("connection_count", 2).twice
-      expect_gauge("memory_usage", 1024).twice
-      expect_gauge("memory_usage_rss", 512).twice
-      expect_gauge("job_count", 5, :status => :processed) # Gauge delta
-      expect_gauge("job_count", 3, :status => :failed) # Gauge delta
-      expect_gauge("job_count", 12, :status => :retry_queue).twice
-      expect_gauge("job_count", 2, :status => :died) # Gauge delta
-      expect_gauge("job_count", 14, :status => :scheduled).twice
-      expect_gauge("job_count", 15, :status => :enqueued).twice
-      expect_gauge("queue_length", 10, :queue => "default").twice
-      expect_gauge("queue_latency", 12_000, :queue => "default").twice
-      expect_gauge("queue_length", 1, :queue => "critical").twice
-      expect_gauge("queue_latency", 2_000, :queue => "critical").twice
-      # Call probe twice so we can calculate the delta for some gauge values
-      probe.call
-      probe.call
-    end
-
-    context "when `redis_info` is not defined" do
-      before do
-        allow(Sidekiq).to receive(:respond_to?).with(:redis_info).and_return(false)
+      it "logs used hostname on call once" do
+        log = capture_logs { probe.call }
+        expect(log).to contains_log(
+          :debug,
+          %(Sidekiq probe: Using Redis server hostname "localhost" as hostname)
+        )
+        log = capture_logs { probe.call }
+        # Match more logs with incompelete message
+        expect(log).to_not contains_log(:debug, %(Sidekiq probe: ))
       end
 
-      it "does not collect redis metrics" do
-        expect_gauge("connection_count", 2).never
-        expect_gauge("memory_usage", 1024).never
-        expect_gauge("memory_usage_rss", 512).never
+      it "collects custom metrics" do
+        expect_gauge("worker_count", 24).twice
+        expect_gauge("process_count", 25).twice
+        expect_gauge("connection_count", 2).twice
+        expect_gauge("memory_usage", 1024).twice
+        expect_gauge("memory_usage_rss", 512).twice
+        expect_gauge("job_count", 5, :status => :processed) # Gauge delta
+        expect_gauge("job_count", 3, :status => :failed) # Gauge delta
+        expect_gauge("job_count", 12, :status => :retry_queue).twice
+        expect_gauge("job_count", 2, :status => :died) # Gauge delta
+        expect_gauge("job_count", 14, :status => :scheduled).twice
+        expect_gauge("job_count", 15, :status => :enqueued).twice
+        expect_gauge("queue_length", 10, :queue => "default").twice
+        expect_gauge("queue_latency", 12_000, :queue => "default").twice
+        expect_gauge("queue_length", 1, :queue => "critical").twice
+        expect_gauge("queue_latency", 2_000, :queue => "critical").twice
+        # Call probe twice so we can calculate the delta for some gauge values
         probe.call
+        probe.call
+      end
+    end
+
+    context "with Sidekiq 6" do
+      before { with_sidekiq6! }
+
+      it "logs used hostname on call once" do
+        log = capture_logs { probe.call }
+        expect(log).to contains_log(
+          :debug,
+          %(Sidekiq probe: Using Redis server hostname "localhost" as hostname)
+        )
+        log = capture_logs { probe.call }
+        # Match more logs with incompelete message
+        expect(log).to_not contains_log(:debug, %(Sidekiq probe: ))
+      end
+
+      it "collects custom metrics" do
+        expect_gauge("worker_count", 24).twice
+        expect_gauge("process_count", 25).twice
+        expect_gauge("connection_count", 2).twice
+        expect_gauge("memory_usage", 1024).twice
+        expect_gauge("memory_usage_rss", 512).twice
+        expect_gauge("job_count", 5, :status => :processed) # Gauge delta
+        expect_gauge("job_count", 3, :status => :failed) # Gauge delta
+        expect_gauge("job_count", 12, :status => :retry_queue).twice
+        expect_gauge("job_count", 2, :status => :died) # Gauge delta
+        expect_gauge("job_count", 14, :status => :scheduled).twice
+        expect_gauge("job_count", 15, :status => :enqueued).twice
+        expect_gauge("queue_length", 10, :queue => "default").twice
+        expect_gauge("queue_latency", 12_000, :queue => "default").twice
+        expect_gauge("queue_length", 1, :queue => "critical").twice
+        expect_gauge("queue_latency", 2_000, :queue => "critical").twice
+        # Call probe twice so we can calculate the delta for some gauge values
+        probe.call
+        probe.call
+      end
+
+      context "when Sidekiq `redis_info` is not defined" do
+        before do
+          allow(Sidekiq).to receive(:respond_to?).with(:redis_info).and_return(false)
+        end
+
+        it "does not collect redis metrics" do
+          expect_gauge("connection_count", 2).never
+          expect_gauge("memory_usage", 1024).never
+          expect_gauge("memory_usage_rss", 512).never
+          probe.call
+        end
       end
     end
 
@@ -179,6 +281,8 @@ describe Appsignal::Probes::SidekiqProbe do
       let(:probe) { described_class.new(:hostname => redis_hostname) }
 
       it "uses the redis hostname for the hostname tag" do
+        with_sidekiq7! # Version irrelevant, but requires any version
+
         allow(Appsignal).to receive(:set_gauge).and_call_original
         log = capture_logs { probe }
         expect(log).to contains_log(

